registry/client_factory.go

```diff
@@ -40,7 +45,26 @@ func (t *logging) RoundTrip(req *http.Request) (*http.Response, error) {
 }
 
 func (f *RemoteClientFactory) ClientFor(repo image.CanonicalName, creds Credentials) (Client, error) {
-	tx := f.Limiters.RoundTripper(http.DefaultTransport, repo.Domain)
+	insecure := false
+	for _, h := range f.InsecureHosts {
+		if repo.Domain == h {
+			insecure = true
+			break
+		}
+	}
```

Comment:
The size of this function is getting out of hand. I would suggest factoring out the challenge obtention (or part of it at least)